ln-gateway/src/lib.rs

```diff
@@ -222,19 +222,32 @@ impl LnGateway {
         payment_params: &PaymentParameters,
         mut rng: impl RngCore + CryptoRng,
     ) -> Result<[u8; 32]> {
-        let (out_point, _) = self
+        let (out_point, contract_id) = self
             .federation_client
             .buy_preimage_offer(
                 &payment_params.payment_hash,
                 &payment_params.invoice_amount,
                 &mut rng,
             )
             .await?;
-        let preimage = self
+
+        match self
             .federation_client
             .await_preimage_decryption(out_point)
-            .await?;
-        Ok(preimage.0.serialize())
+            .await
+        {
+            Ok(preimage) => {
+                debug!("Decrypted preimage {:?}", preimage);
+                Ok(preimage.0.serialize())
+            }
+            Err(e) => {
+                warn!("Failed to decrypt preimage. Now requesting a refund: {}", e);
+                self.federation_client
+                    .refund_incoming_contract(contract_id, rng)
+                    .await?;
```

Comment:
returning an error from `refund_incoming_contract` contract could be confusing but I don't know what to do about that

Reply:
Maybe we can map the error into something with more context?